napari/_qt/qt_dims.py

```diff
@@ -415,16 +278,16 @@ def play(
             reverse.  fps == 0 will stop the animation. The view is not
             guaranteed to keep up with the requested fps, and may drop frames
             at higher fps.
-        frame_range: tuple | list
-            If specified, will constrain animation to loop [first, last] frames
-        playback_mode: str
+        loop_mode: int
```

Comment:
I agree with @royerloic I'd prefer to use our StringEnum class here where you can pass in strings, `play_once`, `loop_continuous`, `loop_back_and_forth` which get converted to their StringEnums - like this StringEnum https://github.com/napari/napari/blob/c28f7316fa8e68c2ba75ee99a5ae3e075833fa38/napari/layers/base/_constants.py#L6 and this conversion https://github.com/napari/napari/blob/c28f7316fa8e68c2ba75ee99a5ae3e075833fa38/napari/layers/base/base.py#L120

It's more consistent with the approach we have taken in the rest of the repo, and I see now major reason to deviate from it here

Reply:
Also for populating the combo box, see how it is done here, iterating of `str()` calls to the Enum - https://github.com/napari/napari/blob/c28f7316fa8e68c2ba75ee99a5ae3e075833fa38/napari/_qt/layers/qt_base_layer.py#L35

it all should be pretty transferable to this use case, with no need to go to `int`.

also feel as you noted below, maybe `once`, `loop`, and `back_and_forth` are better string names

Reply:
I was initially thinking that i liked this... but as I'm implementing this, I find myself having to jump through tons of hoops and boilerplate.  More imports (and new files to avoid circular imports), more lines, less readable code, more opaque signal connections, possibility for bugs (if you forget to convert between enum and string types), and it's still unclear to me exactly what I'm gaining in terms of robustness or even in terms of clarity for the user (that I couldn't get by simply converting one of those strings to an int in `QtDims.play()` which is the only front-facing API where anyone would change these things).  

This really feels like an anti-pattern to me... Other than being consistent with the rest of the repo, what's the reasoning here?

Reply:
I pushed a commit so you can evaluate it... but I still think I have a preference for reverting it.  Another thing I ran into is that `Qt.Signal()` actually threw a segmentation fault when I declared it using the custom StringEnum type... which forces me to use `Signal(str)` and then make sure to convert back and forth between the custom type and `str` wherever that signal is connected.

Reply:
I'd put the Enum definition in a new file `_constants.py` which should avoid any circular import problems. I guess it's nice then to have a single place where constants like this are defined and can be tracked, edited independent of one particular look up table usage, and I prefer the api of `QtDimsSlider` now where mode is the enum instead on an integer which I'm unlikely to remember.

Probably the main difference here vs other places in the code is the `Signal(str)` which means you have an additional conversion. Maybe @jni who is generally more conservative about adding enums want's to weigh in.

Reply:
yeah, I certainly understand wanting to allow the public api to accept strings rather than hard-to-remember ints.  That's easily done without enums, and one way or another we'll definitely have that be the final public API in the `QtDims.play` method.

For me the enum makes the most sense if a) we can imagine wanting to add additional choices in the future and b) we need those updates to be reflected in many places in the code.  In that case an enum lets you make the change in one place.  Here, though it's hard to think of what another `loop_mode` would even be, and it's also not reflected in multiple places throughout the code (it's basically just an argument to the `play` function and the comboBox).  Given the additional difficulty of using a custom enum type in signals and slots, I would probably be further inclined to immediately cast that enum to a string, which more or less means we've created an enum simply to immediately cast to a string after the play function is called.

All that said, I'm happy to just go with whatever the group decision is... so let's give @jni and others time to weigh in and then I'll implement whatever the general opinion is.

Reply:
1) it definitely shouldn't be an int! That's yucky syntax from C.

2) it *could* be a string. From an API and community expectation perspective, that is Pythonic. But,

3) my rationale here and in all the other places where we use enums is: syntactically, an enum is the right structure: the loop mode can be one of several options. A string provides endless possibilities — and endless *erroneous* possibilities. From a type validation perspective, a string is almost completely uninformative in this case. Yes, enums feel unpythonic, but I think that's only because they are a very recent addition: they appeared in Python 3.4. I think in the future, now that enums are in the standard library and that typing is becoming a much bigger part of Python (type annotations, jits, etc), enums will become more common.

For me, really the thing that convinces me to use enums is that they are *the* data type to represent a value from a defined set of options. ints and strings are both kludges to hack around an incomplete language/typing spec. Now that we *have* enums, we should use them when there are multiple, non-numeric, non-boolean options.

Reply:
OK, I'll leave it as is, and just clean it up a bit.